v2/api/keyvault/customizations/vault_extensions.go

```diff
@@ -0,0 +1,376 @@
+		mode := ex.handleCreateOrRecover(ctx, kv, armClient, kubeClient, resolver, log)
+		createMode = &mode
+	}
+
+	if strings.EqualFold(*kv.Spec.Properties.CreateMode, CreateMode_PurgeThenCreate) {
+		err := ex.handlePurgeThenCreate(ctx, kv, armClient, kubeClient, resolver, log)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error purging soft-deleted KeyVault")
+		}
+
+		createMode = to.Ptr(CreateMode_Default)
+	}
+
+	if createMode != nil {
+		// Modify the payload as necessary
+		spec := armObj.Spec()
+		var v string = string(*createMode)
+		err := reflecthelpers.SetProperty(spec, "Properties.CreateMode", &v)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error setting CreateMode to %s", v)
+		}
+	}
+
+	return armObj, nil
+}
+
+func (ex *VaultExtension) handleCreateOrRecover(
+	ctx context.Context,
+	kv *keyvault.Vault,
+	armClient *genericarmclient.GenericClient,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) string {
+	exists, err := ex.checkForExistenceOfDeletedKeyVault(ctx, kv, armClient, kubeClient, resolver, log)
+	if err != nil {
+		// If we can't detect, assume best case and proceed
+		log.Error(err, "error checking for existence of soft-deleted KeyVault")
+		return CreateMode_Default
+	}
+
+	result := CreateMode_Default
+	if exists {
+		result = CreateMode_Recover
+	}
+
+	log.Info(
+		"KeyVault reconciliation requested CreateOrRecover",
+		"KeyVault", kv.Name,
+		"Soft-deleted-KeyVault-exists", exists,
+		"createMode", result)
+
+	return result
+}
+
+func (ex *VaultExtension) handlePurgeThenCreate(
+	ctx context.Context,
+	kv *keyvault.Vault,
+	armClient *genericarmclient.GenericClient,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) error {
+	// Find out whether a soft-deleted KeyVault with the same name exists
+	exists, err := ex.checkForExistenceOfDeletedKeyVault(ctx, kv, armClient, kubeClient, resolver, log)
+	if err != nil {
+		// Couldn't determine whether a soft-deleted keyvault exists, assume it doesn't
+
+		log.Error(err, "error checking for existence of soft-deleted KeyVault")
+		return nil
+	}
+
+	log.Info(
+		"KeyVault reconciliation requested PurgeThenCreate",
+		"KeyVault", kv.Name,
+		"Deleted KeyVault exists", exists)
+
+	if exists {
+		// if a soft-deleted KeyVault exists, we need to purge it before we can create a new one
+		req, err := ex.createPurgeRequest(ctx, kv, armClient, kubeClient, resolver, log)
+		if err != nil {
+			return errors.Wrapf(err, "error creating purge request for KeyVault %s", kv.Name)
+		}
+
+		_, err = armClient.Post(ctx, req)
+		if err != nil {
+			return errors.Wrapf(err, "error purging KeyVault %s", kv.Name)
+		}
+	}
+
+	return nil
+}
+
+// checkForExistenceOfDeletedKeyVault checks to see whether there's a soft deleted KeyVault with the same name.
+// This might be true if another party has deleted the KeyVault, even if we previously created it
+func (ex *VaultExtension) checkForExistenceOfDeletedKeyVault(
```

Comment:
Can use [GetDeleted](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/keyvault/armkeyvault/vaults_client.go#L276) here from `azure-sdk-for-go` ?